### PR TITLE
flatten statements helper

### PR DESCRIPTION
**Pull request by Kroppeb**

### CODE REVIEW

File: src/org/jetbrains/java/decompiler/code/cfg/ControlFlowGraph.java [hunk 1]
```diff
@@ -854,6 +856,12 @@ public Set<BasicBlock> getFinallyExits() {
     return finallyExits;
   }
 
+  /*
```
Comment (SuperCoder7979) [inline src/org/jetbrains/java/decompiler/code/cfg/ControlFlowGraph.java:859]: Leftover comment blocks

File: src/org/jetbrains/java/decompiler/modules/decompiler/decompose/DomHelper.java [hunk 2]
```diff
@@ -26,7 +24,8 @@ public final class DomHelper implements GraphParser {
 
   @Override
   public RootStatement createStatement(ControlFlowGraph graph, StructMethod mt) {
-    return parseGraph(graph, mt);
+    // TODO: I'm passing in iteration 0 now, but I probably wan to pass this in from the caller
```
Comment (SuperCoder7979) [inline src/org/jetbrains/java/decompiler/modules/decompiler/decompose/DomHelper.java:27]: Remove debug comments in class

File: src/org/jetbrains/java/decompiler/modules/decompiler/decompose/DomHelper.java [hunk 3]
```diff
@@ -486,22 +493,87 @@ private static boolean processStatement(Statement general, RootStatement root, H
...
+      }
+    }
+    return null;
+  }
+
+  private static Statement findGeneralStatement(
+    Statement stat,
+    boolean forceall,
+    HashMap<Integer, Set<Integer>> mapExtPost
+    /* Set<FinallyProcessor.FinallyInformation> finallyInfos*/) {
```
Comment (SuperCoder7979) [inline src/org/jetbrains/java/decompiler/modules/decompiler/decompose/DomHelper.java:524]: Remove comments and unused `findStatementByBlock` function

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java [hunk 4]
```diff
@@ -80,722 +97,553 @@ private DirectNode createDirectNode(Statement stat, DirectNodeType type) {
...
+          // add implicit else edge
+          this.addEdge(outNode, stat.getFirstDirectSuccessor());
+        }
+
+        // Adds an edge from the last if statement to the current if statement,
+        // if the current if statement's head statement has no predecessor
+        // This was made to mask a failure in EliminateLoopsHelper and isn't used currently
+        // (over the current test set) but could theoretically still happen!
+        // TODO: what?
+        // TODO: use java code gen to generate a test for this?
```
Comment (SuperCoder7979) [inline src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:471]: Just remove this code block, I don't think it's needed anymore

File: src/org/jetbrains/java/decompiler/util/DotExporter.java [hunk 5]
```diff
@@ -49,7 +49,11 @@ public class DotExporter {
   // Statements with no successors or predecessors (but still contained in the tree) will be in a subgraph titled "Isolated statements".
   // Statements that aren't found will be circular, and will have a message stating so.
   // Nodes with green borders are the canonical exit of method, but these may not always be emitted.
+
   private static String statToDot(Statement stat, String name) {
+    return statToDot(stat, name, null);
+  }
```
Comment (SuperCoder7979) [inline src/org/jetbrains/java/decompiler/util/DotExporter.java:55]: Whitespace

**Review by SuperCoder7979 — CHANGES_REQUESTED** [review]
Very helpful changes, I have a few code style and debug change requests

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java (hunk 4 above)
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:471]: ok

File: src/org/jetbrains/java/decompiler/util/DotExporter.java (hunk 5 above)
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/util/DotExporter.java:55]: Ok

**Review by SuperCoder7979 — CHANGES_REQUESTED** [review]

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java [hunk 6]
```diff
@@ -80,722 +97,527 @@ private DirectNode createDirectNode(Statement stat, DirectNodeType type) {
...
 
-      Deque<StackEntry> stackFinally = statEntry.stackFinally;
-      int statementBreakIndex = statEntry.statementIndex;
+          // Look if this basic block is the successor of a sequence,
+          // and connect the sequence to the block if so
+          // TODO: should this be done in the sequence handling instead?
+          // TODO: what is this for?
+          // TODO: why is it adding ad edge from the start of the sequence to the block?
+          // TODO: why is it using the basic head id, instead of the sequence id?
+          // TODO: disabling this seems to not cause any validation errors, or changes in output,
```
[the remark this answers is not in the document]
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:138]: I don't know?

File: src/org/jetbrains/java/decompiler/modules/decompiler/decompose/DomHelper.java (hunk 2 above)
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/modules/decompiler/decompose/DomHelper.java:27]: remove or change it?

Comment (SuperCoder7979) [reply src/org/jetbrains/java/decompiler/modules/decompiler/decompose/DomHelper.java:27]: Remove it

File: src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java [hunk 7]
```diff
@@ -539,15 +538,20 @@ private boolean verifyFinallyEx(ControlFlowGraph graph, CatchAllStatement fstat,
...
 
       if (entry.getValue()) {
         removeExceptionInstructionsEx(last, 2, finallytype);
         graph.getFinallyExits().add(last);
+        ValidationHelper.assertTrue(trueExit == null, "More than one true exit");
+        trueExit = last;
       }
     }
 
+    // graph.getFinallyInfos().add(new FinallyInformation(fstat.getHandler().getBasichead().getBlock(), trueExit, sideExits));
```
Comment (SuperCoder7979) [inline src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java:553]: debug

**Review by SuperCoder7979 — CHANGES_REQUESTED** [review]
one more

**Pull request merged**